File: 09_Remote_Procedures/demo/helloworld/java/src/main/java/Main.java
package greeter;

import io.grpc.Server;
import io.grpc.ServerBuilder;
import io.grpc.stub.StreamObserver;
import static greeter.Greeter.*;

public class Main {

    public static void main(String[] args) throws Exception {
        Server server = ServerBuilder.forPort(50052).addService(new SayerImpl()).build();
        server.start();
        System.out.println("Server started on :50052");
        server.awaitTermination();
    }

    static class SayerImpl extends SayerGrpc.SayerImplBase {
        @Override
        public void sayHello(HelloRequest request, StreamObserver<HelloReply> responseObserver) {
            try { Thread.sleep(0); } catch (InterruptedException e) { Thread.currentThread().interrupt(); }
            HelloReply reply = HelloReply.newBuilder().setMessage("Hello " + request.getName() + "!").build();
            responseObserver.onNext(reply);
            responseObserver.onCompleted();
        }
    }
}

File: 09_Remote_Procedures/demo/helloworld/go/main.go
package main

import (
	"context"
	"log"
	"net"
	"time"

	"google.golang.org/grpc"

	pb "demo/greeter"
)

type askerServer struct {
	pb.UnimplementedAskerServer
}

func (s *askerServer) AskHello(context context.Context, request *pb.HelloRequest) (*pb.HelloReply, error) {
	time.Sleep(3000 * time.Millisecond)
	return &pb.HelloReply{
		Message: "Hello " + request.GetName() + "?",
	}, nil
}

func main() {
	listener, _ := net.Listen("tcp", ":50051")
	grpcServer := grpc.NewServer()
	pb.RegisterAskerServer(grpcServer, &askerServer{})
	log.Println("Server started on :50051")
	grpcServer.Serve(listener)
}

